rent/src/lib.rs

```diff
@@ -0,0 +1,1489 @@
+
+			// Remove contract.
+			Contracts::<T>::remove(nft_id);
+
+			// Deposit event.
+			let event = Event::ContractRevoked { nft_id, revoked_by: who };
+			Self::deposit_event(event);
+
+			Ok(().into())
+		}
+
+		/// Rent an nft if contract exist, makes an offer if it's manual acceptance.
+		#[pallet::weight((
+			{
+				let (s, t) = Contracts::<T>::get(nft_id)
+					.map_or_else(|| (0, 0), |c| {
+						let mut queues = Queues::<T>::get();
+						let available_size = queues.available_queue.size();
+						match c.acceptance_type {
+							AcceptanceType::AutoAcceptance(_) => {
+								match c.duration {
+									Duration::Subscription(_, _) => (available_size, queues.subscription_queue.size()),
+									Duration::Fixed(_) =>  (available_size, queues.fixed_queue.size()),
+									Duration::Infinite => (available_size, 0)
+								}
+							},
+							AcceptanceType::ManualAcceptance(_) => (0, 0)
+						}
+					});
+				T::WeightInfo::rent(s as u32, t as u32)
+			},
+			DispatchClass::Normal
+		))]
+		pub fn rent(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			Contracts::<T>::try_mutate(nft_id, |x| -> DispatchResult {
+				let contract = x.as_mut().ok_or(Error::<T>::ContractNotFound)?;
+				ensure!(contract.renter != who, Error::<T>::CannotRentOwnContract);
+
+				Self::ensure_enough_for_rent(
+					&who,
+					&contract.rent_fee,
+					&contract.rentee_cancellation_fee,
+				)?;
+
+				match contract.acceptance_type.clone() {
+					AcceptanceType::ManualAcceptance(accounts) => {
+						if let Some(accounts) = accounts {
+							ensure!(accounts.contains(&who), Error::<T>::NotAuthorizedForRent);
+						}
+
+						Self::insert_offer(nft_id, who.clone())?;
+
+						// Deposit event.
+						let event = Event::ContractOfferCreated { nft_id, rentee: who };
+						Self::deposit_event(event);
+					},
+					AcceptanceType::AutoAcceptance(accounts) => {
+						if let Some(accounts) = accounts {
+							ensure!(accounts.contains(&who), Error::<T>::NotAuthorizedForRent);
+						}
+
+						let mut queues = Queues::<T>::get();
+						let now = <frame_system::Pallet<T>>::block_number();
+						Self::take_rent_fee(&who, contract.renter.clone(), &contract.rent_fee)?;
+						Self::take_cancellation_fee(&who, &contract.rentee_cancellation_fee)?;
+
+						if let Some(expiration_block) =
+							Self::get_expiration_block(&contract.duration, now)
+						{
+							queues
+								.insert_in_queue(nft_id, &contract.duration, expiration_block)
+								.map_err(|_| Error::<T>::MaxSimultaneousContractReached)?;
+						}
+
+						queues.remove_from_available_queue(nft_id);
+						Queues::<T>::set(queues);
+
+						Offers::<T>::remove(nft_id);
+
+						contract.terms_accepted = true;
+						contract.rentee = Some(who.clone());
+						contract.has_started = true;
+						contract.start_block = Some(<frame_system::Pallet<T>>::block_number());
+
+						Contracts::<T>::insert(nft_id, contract);
```

Comment:
This seems like a bug.  You are mutate the contract under nft id X but at the same time you are inserting a new contract with the same nft id?